gateway/src/main/scala/filodb/gateway/conversion/InputRecord.scala

```diff
@@ -41,52 +44,79 @@ trait InputRecord {
                                  dataset: Dataset,
                                  timestamp: Long,
                                  value: Double) extends InputRecord {
   import collection.JavaConverters._
 
-  val shardKeys = dataset.options.shardKeyColumns.toSet
-  val scalaKVs = tags.toSeq
-  val originalKVs = new java.util.ArrayList(scalaKVs.asJava)
-  val forShardKVs = scalaKVs.map { case (k, v) =>
-                      val trimmedVal = RecordBuilder.trimShardColumn(dataset, k, v)
-                      (k, trimmedVal)
-                    }
-  val kvsForShardCalc = new java.util.ArrayList(forShardKVs.asJava)
+  val trimmedMetric = RecordBuilder.trimShardColumn(dataset, dataset.options.metricColumn, metric)
+  val javaTags = new java.util.ArrayList(tags.toSeq.asJava)
 
   // Get hashes and sort tags of the keys/values for shard calculation
-  val hashes = RecordBuilder.sortAndComputeHashes(kvsForShardCalc)
+  val hashes = RecordBuilder.sortAndComputeHashes(javaTags)
 
-  final def shardKeyHash: Int = RecordBuilder.combineHashIncluding(kvsForShardCalc, hashes, shardKeys)
-                                             .getOrElse(PrometheusInputRecord.DefaultShardHash)
-  final def partitionKeyHash: Int = RecordBuilder.combineHashExcluding(kvsForShardCalc, hashes, shardKeys)
+  final def shardKeyHash: Int = RecordBuilder.shardKeyHash(nonMetricShardValues, trimmedMetric)
+  final def partitionKeyHash: Int = RecordBuilder.combineHashExcluding(javaTags, hashes,
+                                                    dataset.options.ignorePartKeyHashTags)
 
-  final def nonMetricShardValues: Seq[String] = dataset.options.nonMetricShardColumns.map(tags)
-  final def getMetric: String = tags(dataset.options.metricColumn)
+  val nonMetricShardValues: Seq[String] = dataset.options.nonMetricShardColumns.flatMap(tags.get)
+  final def getMetric: String = metric
 
   final def addToBuilder(builder: RecordBuilder): Unit = {
-    originalKVs.sort(RecordBuilder.stringPairComparator)
     builder.startNewRecord()
     builder.addLong(timestamp)
     builder.addDouble(value)
-    builder.addSortedPairsAsMap(originalKVs, hashes)
+    builder.startMap()
+    val metricBytes = metric.getBytes
+    builder.addMapKeyValueHash(dataset.options.metricBytes, dataset.options.metricHash,
+                               metricBytes, 0, metricBytes.size)
+    for { i <- 0 until javaTags.size optimized } {
+      val (k, v) = javaTags.get(i)
+      builder.addMapKeyValue(k.getBytes, v.getBytes)
+      builder.updatePartitionHash(hashes(i))
+    }
+    builder.endMap()
+
     builder.endRecord()
   }
 }
 
 object PrometheusInputRecord {
   val DefaultShardHash = -1
 
-  // Create PrometheusInputRecord from a TimeSeries protobuf object
-  def apply(tsProto: TimeSeries, dataset: Dataset): PrometheusInputRecord = {
+  // Create PrometheusInputRecords from a TimeSeries protobuf object
+  def apply(tsProto: TimeSeries, dataset: Dataset): Seq[PrometheusInputRecord] = {
```

Comment:
It appears this method along with transform tags is doing many iterations of tags. The scala map/filter constructs are not helping. Can we just do one (or as few) iterations of the tags and calculate everything in a mutable way ? That would be slightly more efficient and will create less garbage as well. Helps since it is in the ingestion path for each and every sample.

Reply:
@vishramachandran let's leave this for later if this is really needed.   There is really no point to put any more effort into the Prometheus protobuf format, as we are moving away from it long term.  The code in here already speeds up the process by 15% actually.